reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/ConcatMapTest.kt

```diff
@@ -0,0 +1,214 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.testutils.TestObservable
+import com.badoo.reaktive.testutils.TestObservableObserver
+import com.badoo.reaktive.testutils.dispose
+import com.badoo.reaktive.testutils.hasOnNext
+import com.badoo.reaktive.testutils.hasSubscribers
+import com.badoo.reaktive.testutils.isCompleted
+import com.badoo.reaktive.testutils.isError
+import com.badoo.reaktive.testutils.test
+import com.badoo.reaktive.testutils.values
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFalse
+import kotlin.test.assertTrue
+
+class ConcatMapTest :  UpstreamDownstreamGenericTests by UpstreamDownstreamGenericTests<Unit>({ concatMap { TestObservable<Int>() } }){
+
+    private val upstream = TestObservable<Int?>()
+
+    @Test
+    fun subscribes_to_upstream() {
+        testConcatMap { TestObservable() }
+
+        assertTrue(upstream.hasSubscribers)
+    }
+
+    @Test
+    fun subscribes_to_first_inner_source() {
+        val inner = TestObservable<String>()
+        testConcatMap { inner }
+
+        upstream.onNext(0)
+
+        assertTrue(inner.hasSubscribers)
+    }
+
+    @Test
+    fun does_not_subscribe_to_second_inner_source_WHEN_first_inner_source_is_not_finished() {
+        val inners = innerSources(2)
+        testConcatMap(inners)
+
+        upstream.onNext(0)
+        upstream.onNext(1)
+
+        assertFalse(inners[1].hasSubscribers)
+    }
+
+    @Test
+    fun subscribes_to_second_source_WHEN_upstream_emitted_second_value_and_first_source_is_finished() {
+        val inners = innerSources(2)
+        testConcatMap(inners)
+
+        upstream.onNext(0)
+        upstream.onNext(1)
+        inners[0].onComplete()
+
+        assertTrue(inners[1].hasSubscribers)
+    }
+
+    @Test
+    fun subscribes_to_second_source_WHEN_first_source_is_finished_and_upstream_emitted_second_value() {
+        val inners = innerSources(2)
+        testConcatMap(inners)
+
+        upstream.onNext(0)
+        inners[0].onComplete()
+        upstream.onNext(1)
+
+        assertTrue(inners[1].hasSubscribers)
+    }
+
+    @Test
+    fun produces_values_in_correct_order() {
+        val inners = innerSources(3)
+        val observer = testConcatMap() { inners[it ?: 0] }
+
+        upstream.onNext(null)
```

Comment:
Would be nice if we had some way of making this test more legible. However I'm out of ideas.